File: src/main/java/cope/nebula/client/feature/module/render/Nametags.java
package cope.nebula.client.feature.module.render;

import com.mojang.realmsclient.gui.ChatFormatting;
import cope.nebula.asm.duck.IEntityPlayer;
import cope.nebula.client.feature.module.Module;
import cope.nebula.client.feature.module.ModuleCategory;
import cope.nebula.client.value.Value;
import cope.nebula.util.renderer.FontUtil;
import cope.nebula.util.renderer.RenderUtil;
import cope.nebula.util.world.entity.EntityUtil;
import cope.nebula.util.world.entity.player.rotation.AngleUtil;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.client.renderer.entity.RenderManager;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumHand;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Nametags extends Module {
    public static Nametags INSTANCE;

    public Nametags() {
        super("Nametags", ModuleCategory.RENDER, "Renders nametags above players heads");
        INSTANCE = this;
    }

    public static final Value<Boolean> held = new Value<>("Held", false);

    public static final Value<Boolean> armor = new Value<>("Armor", true);
    public static final Value<Boolean> reversed = new Value<>(armor, "Reversed", true);
    public static final Value<Boolean> enchants = new Value<>(armor, "Enchants", true);
    public static final Value<Boolean> percent = new Value<>(armor, "Percent", true);

    public static final Value<Boolean> health = new Value<>("Health", true);
    public static final Value<Boolean> ping = new Value<>("Ping", true);
    public static final Value<Boolean> pops = new Value<>("Pops", true);

    public static final Value<Double> scaling = new Value<>("Scaling", 2.0, 1.0, 5.0);

    @Override
    public void onRender3d() {
        if (mc.getRenderViewEntity() != null) {
            for (EntityPlayer player : mc.world.playerEntities) {
                if (player.equals(mc.player) || player.isDead) {
                    continue;
                }

                RenderManager renderManager = mc.getRenderManager();

                double x = AngleUtil.interpolate(player.posX, player.prevPosX) - renderManager.viewerPosX;
                double y = AngleUtil.interpolate(player.posY, player.prevPosY) - renderManager.viewerPosY;
                double z = AngleUtil.interpolate(player.posZ, player.prevPosZ) - renderManager.viewerPosZ;

                renderNametag(player, x, y, z);
            }
        }
    }

    private void renderNametag(EntityPlayer player, double x, double y, double z) {
        double offset = player.isSneaking() ? 0.5 : 0.7;

        RenderManager renderManager = mc.getRenderManager();

        double distance = mc.getRenderViewEntity().getDistance(x + renderManager.viewerPosX, y + renderManager.viewerPosY, z + renderManager.viewerPosZ);
        double scale = Math.max((scaling.getValue() / 10.0) * 4.0, (scaling.getValue() / 10.0) * distance) / 50.0;

        GlStateManager.pushMatrix();
        RenderHelper.enableGUIStandardItemLighting();

        GlStateManager.enablePolygonOffset();
        GlStateManager.doPolygonOffset(1.0f, -1500000.0f);

        GlStateManager.disableLighting();
        GlStateManager.translate(x, 1.4 + y + offset, z);

        GlStateManager.rotate(-renderManager.playerViewY, 0.0f, 1.0f, 0.0f);
        GlStateManager.rotate(renderManager.playerViewX, mc.gameSettings.thirdPersonView == 2 ? -1.0f : 1.0f, 0.0f, 0.0f);

        GlStateManager.scale(-scale, -scale, scale);

        GlStateManager.disableDepth();
        GlStateManager.enableBlend();

        String info = getNametagInfo(player);
        float width = FontUtil.getWidth(info) / 2.0f;

        RenderUtil.drawRectangle(-(width - 4.0), -(FontUtil.getHeight() - 1.0f), (width * 2.0f) + 4.0f, FontUtil.getHeight() + 2.0f, 0x70000000);
        FontUtil.drawString(info, -(width - 6.0f), -(FontUtil.getHeight() - 2.0f), -1);

        int xOffset = -24 / 2 * player.inventory.armorInventory.size();

        if (held.getValue()) {
            ItemStack stack = player.getHeldItemMainhand();
            if (!stack.isEmpty()) {
                renderItem(stack, xOffset);
            }

            xOffset += 16;
        }

        if (armor.getValue()) {
            List<ItemStack> items = new ArrayList<>(player.inventory.armorInventory);
            if (reversed.getValue()) {
                Collections.reverse(items);
            }

            for (ItemStack stack : items) {
                if (!stack.isEmpty()) {
                    renderItem(stack, xOffset);
                    xOffset += 16;
                }
            }
        }

        if (held.getValue()) {
            ItemStack stack = player.getHeldItemOffhand();
            if (!stack.isEmpty()) {
                renderItem(stack, xOffset);
                xOffset += 16;
            }
        }

        GlStateManager.enableDepth();
        GlStateManager.depthMask(true);
        GlStateManager.enableLighting();
        GlStateManager.disableBlend();

        GlStateManager.doPolygonOffset(1.0f, 1500000.0f);
        GlStateManager.disablePolygonOffset();

        GlStateManager.popMatrix();
    }

    private void renderItem(ItemStack stack, int x) {
        GlStateManager.pushMatrix();
        GlStateManager.depthMask(true);
        RenderHelper.enableStandardItemLighting();
        mc.getRenderItem().zLevel = -150.0f;
        GlStateManager.disableAlpha();
        GlStateManager.enableDepth();
        GlStateManager.disableCull();

        mc.getRenderItem().renderItemAndEffectIntoGUI(stack, x, -26);
        mc.getRenderItem().renderItemOverlays(mc.fontRenderer, stack, x, -26);

        mc.getRenderItem().zLevel = 0.0f;
        RenderHelper.disableStandardItemLighting();
        GlStateManager.enableCull();
        GlStateManager.enableAlpha();
        GlStateManager.enableDepth();
        GlStateManager.popMatrix();
    }

    private String getNametagInfo(EntityPlayer player) {
        StringBuilder builder = new StringBuilder();
        if (((IEntityPlayer) player).isFriend()) {
            builder.append(ChatFormatting.AQUA);
        } else if (player.isSneaking()) {
            builder.append(ChatFormatting.GOLD);
        }

        builder.append(player.getName());

        if (health.getValue()) {
            builder.append(" ");

            float health = EntityUtil.getHealth(player);

            if (health >= 20.0f) {
                builder.append(ChatFormatting.GREEN);
            } else if (health < 15.0f) {
                builder.append(ChatFormatting.YELLOW);
            } else {
                builder.append(ChatFormatting.RED);
            }

            builder.append(String.format("%.2f", health)).append(ChatFormatting.RESET);
        }

        if (ping.getValue()) {
            builder.append(" ")
                    .append(getNebula().getServerManager().getLatency(player.getUniqueID()))
                    .append("ms");
        }

        if (pops.getValue()) {
            // TODO
        }

        return builder.toString();
    }
}

File: src/main/kotlin/wtf/nebula/client/feature/module/render/Nametags.kt
package wtf.nebula.client.feature.module.render

import com.mojang.realmsclient.gui.ChatFormatting
import me.bush.eventbuskotlin.EventListener
import me.bush.eventbuskotlin.listener
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderHelper
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import wtf.nebula.client.Nebula
import wtf.nebula.client.event.render.RenderWorldEvent
import wtf.nebula.client.feature.module.Module
import wtf.nebula.client.feature.module.ModuleCategory
import wtf.nebula.util.render.RenderUtil
import wtf.nebula.util.rotation.AngleUtil

class Nametags : Module(ModuleCategory.RENDER, "custom nametags and whatnot") {
    val scale by double("Scale", 0.2, 0.1..1.3)
    val armor by bool("Armor", true)
    val reversed by bool("Reversed Armor", true)
    val ping by bool("Ping", true)
    val health by bool("Health", true)
    val held by bool("Held", true)
    val enchantments by bool("Enchantments", true)

    @EventListener
    private val renderWorldListener = listener<RenderWorldEvent> {
        mc.renderViewEntity ?: return@listener

        for (entityPlayer in mc.world.playerEntities) {
            val x = AngleUtil.interpolate(entityPlayer.posX, entityPlayer.lastTickPosX) - mc.renderManager.viewerPosX
            val y = AngleUtil.interpolate(entityPlayer.posY, entityPlayer.lastTickPosY) - mc.renderManager.viewerPosY
            val z = AngleUtil.interpolate(entityPlayer.posZ, entityPlayer.lastTickPosZ) - mc.renderManager.viewerPosZ

            renderNametag(entityPlayer, x, y, z)
        }
    }

    private fun renderNametag(entityPlayer: EntityPlayer, x: Double, y: Double, z: Double) {
        val renderX = mc.renderManager.viewerPosX
        val renderY = mc.renderManager.viewerPosY
        val renderZ = mc.renderManager.viewerPosZ

        val distance = mc.renderViewEntity!!.getDistance(x + renderX, y + renderY, z + renderZ)
        val scale = scale * distance.coerceAtLeast(4.0) / 50.0

        GlStateManager.pushMatrix()
        RenderHelper.enableGUIStandardItemLighting()

        GlStateManager.enablePolygonOffset()
        GlStateManager.doPolygonOffset(1.0f, -1500000.0f)

        GlStateManager.disableLighting()

        GlStateManager.translate(x, y + entityPlayer.height + 0.25, z)

        GlStateManager.rotate(-mc.renderManager.playerViewY, 0.0f, 1.0f, 0.0f)
        GlStateManager.rotate(mc.renderManager.playerViewX, if (mc.gameSettings.thirdPersonView == 2) -1.0f else 1.0f, 0.0f, 0.0f)
        GlStateManager.scale(-scale, -scale, scale)

        GlStateManager.disableDepth()
        GlStateManager.enableBlend()

        val info = getNametagInfo(entityPlayer)

        val width = mc.fontRenderer.getStringWidth(info) / 2.0f
        RenderUtil.rect(
            -(width - 2.0),
            -(mc.fontRenderer.FONT_HEIGHT - 1.0),
            width * 2.0 + 2.0,
            mc.fontRenderer.FONT_HEIGHT + 2.0,
            0x70000000
        )

        mc.fontRenderer.drawStringWithShadow(
            info,
            -(width - 4.0).toInt().toFloat(),
            -(mc.fontRenderer.FONT_HEIGHT - 2).toFloat(),
            -1
        )

        var xOffset = -24 / 2 * entityPlayer.inventory.armorInventory.size

        if (held) {

            if (!entityPlayer.heldItemMainhand.isEmpty) {
                renderItem(entityPlayer.heldItemMainhand, xOffset)
            }

            xOffset += 16
        }

        if (armor) {
            val armor = entityPlayer.inventory.armorInventory

            for (armorPiece in armor) {
                if (!armorPiece.isEmpty) {
                    renderItem(armorPiece, xOffset)
                    xOffset += 16
                }
            }
        }

        if (held) {
            if (!entityPlayer.heldItemOffhand.isEmpty) {
                renderItem(entityPlayer.heldItemOffhand, xOffset)
            }

            xOffset += 16
        }

        GlStateManager.enableDepth();
        GlStateManager.depthMask(true);

        GlStateManager.enableLighting();
        GlStateManager.enableBlend();

        GlStateManager.doPolygonOffset(1.0f, 1500000.0f);
        GlStateManager.disablePolygonOffset();

        GlStateManager.popMatrix();
    }

    private fun renderItem(stack: ItemStack, x: Int) {
        GlStateManager.pushMatrix()
        GlStateManager.depthMask(true)

        GlStateManager.clear(256)

        RenderHelper.enableStandardItemLighting()

        mc.renderItem.zLevel = -150.0f

        GlStateManager.disableAlpha()
        GlStateManager.enableDepth()
        GlStateManager.disableBlend()
        GlStateManager.disableCull()

        mc.renderItem.renderItemAndEffectIntoGUI(stack, x, -26)
        mc.renderItem.renderItemOverlays(mc.fontRenderer, stack, x, -26)

        mc.renderItem.zLevel = 0.0f

        RenderHelper.disableStandardItemLighting()

        GlStateManager.enableCull()
        GlStateManager.enableBlend()
        GlStateManager.enableAlpha()
        GlStateManager.enableDepth()
        GlStateManager.popMatrix()
    }

    private fun getNametagInfo(entityPlayer: EntityPlayer): String {
        val stringBuilder = StringBuilder()

        if (ping) {
            var latency = 0

            try {
                for (info in mc.player.connection.playerInfoMap) {
                    info ?: continue

                    if (info.gameProfile.name == entityPlayer.gameProfile.name) {
                        latency = info.responseTime
                        break
                    }
                }
            } catch (ignored: Exception) {

            }

            stringBuilder.append(latency.toString()).append("ms").append(" ")
        }

        val friend = Nebula.friendManager.friends[entityPlayer.name]
        if (friend != null) {
            stringBuilder.append(ChatFormatting.AQUA)
        }

        else if (entityPlayer.isSneaking) {
            stringBuilder.append(ChatFormatting.GOLD)
        }

        if (friend != null) {
            stringBuilder.append(friend.alias)
        }

        else {
            stringBuilder.append(entityPlayer.name)
        }

        stringBuilder.append(ChatFormatting.RESET)

        if (health) {
            stringBuilder.append(" ")

            val health: Float = entityPlayer.health + entityPlayer.absorptionAmount

            if (health >= 20.0f) {
                stringBuilder.append(ChatFormatting.GREEN)
            } else if (health < 12.0f) {
                stringBuilder.append(ChatFormatting.RED)
            } else {
                stringBuilder.append(ChatFormatting.GREEN)
            }

            stringBuilder.append(String.format("%.1f", health))
        }

        return stringBuilder.toString()
    }
}